src/libsyntax/ast.rs

```diff
@@ -178,21 +178,47 @@ impl PathParameters {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum GenericAngleBracketedParam {
+    Lifetime(Lifetime),
+    Type(P<Ty>),
+}
+
 /// A path like `Foo<'a, T>`
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug, Default)]
 pub struct AngleBracketedParameterData {
     /// Overall span
     pub span: Span,
-    /// The lifetime parameters for this path segment.
-    pub lifetimes: Vec<Lifetime>,
-    /// The type parameters for this path segment, if present.
-    pub types: Vec<P<Ty>>,
+    /// The parameters for this path segment.
+    pub parameters: Vec<GenericAngleBracketedParam>,
```

Comment:
Doing the same transformation in HIR may be a better idea because we can unify with fn-like generic arguments, but from what I see it still makes things more complex, less convenient and we have to allocate just to visit something.
Can we maybe implement AST/HIR support for const generics first and only then exercise in questionable refactoring?

Reply:
I'm also not a fan of generic `enum GenericParameter` used both in HIR and in `ty`.
There's plenty of opportunities to do this generalization in various places in HIR and ty, but this is not generally done, because... that's not useful, there's no reason to generalize here, let's keep HIR and ty separate.

Reply:
Okay, I guess without `enum GenericParameter` the HIR refactoring is not that bad if we avoid allocation and use the same iterator-based approach to `fn lifetimes`/`fn types` as https://github.com/rust-lang/rust/pull/45930 does.

Reply:
Also the things in this PR are `GenericArgs`, not `GenericParameters` (these already exist) or especially `PathParameters` (it's hard to come up with worse name, they are neither parameters, not apply to a path), good opportunity to rename!

Reply:
I'm going to do a more thorough review but 1. I'm against abstracting `enum` definitions when it doesn't have a drastically positive impact (I've debated this on IRC already) 2. I have high hopes for getting rid of `lifetimes` / `types` methods in most places, and the few remaining places should just do it by hand with exhaustive matches.

Reply:
Oh, also, one of the reasons I'm against over-abstracting `enum`s` is because in some places there are actual common fields (like IDs & names) and it'd be useful to extract those out.

Reply:
In addition to not doing this refactoring on AST, I'd actually prefer to revert the AST part of https://github.com/rust-lang/rust/pull/45930, at least until all kinds of generic parameters are allowed to intermingle (which may never happen). AST is supposed to be close to the source and I don't think it's desirable to make things impossible in source (like `<'a, T, U, 'b>`) representable in it.

Reply:
That might be a worthwhile goal, but didn't the `const` generics RFC impose no order between type and `const` parameters? cc @withoutboats @nikomatsakis 

> Can we maybe implement AST/HIR support for const generics first and only then exercise in questionable refactoring?

FWIW I've been asking (on IRC) for the refactoring *before* landing `const` generics, because I'm *really* worried that certain parts of the compiler which hackily use separate vectors for lifetimes and types will get even worse and make the refactoring more difficult.

Reply:
@petrochenkov: It doesn't really make const generics _closer_ — it's intended to reduce the amount of boilerplate when they're introduced, as well as trigger errors if we forget to handle them in a particular spot.

I dislike the `types()`/`lifetimes()` usages too. I'd like to remove them wherever possible, as @eddyb suggests. I've removed the obvious instances where they were unnecessary, but hopefully there's a lot more clean up that can be done.

The names are effectively all placeholders at this stage — my main aim was to keep them separate, so they were easy to identity. It's very easy to rename them all!

@eddyb: if `lifetimes()`/`types()` can be mostly removed, then the common enum is less useful. But if no-one else likes it, I'll remove it now...

> In addition to not doing this refactoring on AST, I'd actually prefer to revert the AST part of #45930, at least until all kinds of generic parameters are allowed to intermingle (which may never happen).

> That might be a worthwhile goal, but didn't the const generics RFC impose no order between type and const parameters?

Ordering of type and const parameters was left as an open question in the RFC. However, I think it really makes sense to restrict const parameters (at least to _begin_ with) to come after type parameters because of the dependencies it implies:

Types can depend on lifetimes; consts can depend on types (possible in theory, though not part of RFC 2000).

If consts can come before types, to me that implies that types dependent on consts can be specified, and Rust definitely isn't ready for those yet.